File: syllabus-1/day-2/RISK.py
cord = [(-1, 0), (1, 0), (0, -1), (0, 1)]


def isValid(x, y, n, m):
    return (x >= 0 and y >= 0 and x < m and y < n)


def solve():
    (n, m) = map(int, input().split())

    arr = []
    for i in range(n):
        arr.append([i for i in input()])

    res = []
    count = 0
    for i in range(n):
        for j in range(m):
            if(arr[i][j] == '0'):
                continue
            # [i,j]
            q = []
            q.append((j, i))
            arr[i][j] = '0'

            while(len(q) > 0):
                l = len(q)
                while(l > 0):
                    x = q[0][0]
                    y = q[0][1]
                    q.pop(0)

                    count += 1
                    for k in range(4):
                        cx = x + cord[k][0]
                        cy = y + cord[k][1]

                        if(isValid(cx, cy, n, m) and arr[cy][cx] == '1'):
                            arr[cy][cx] = '0'
                            q.append((cx, cy))
                    l -= 1
            res.append(count)
            count = 0
    res.sort(reverse=True)
    # print(res)
    ans = 0
    for i in range(1, len(res), 2):
        ans += res[i]
    print(ans)


t = int(input())
while(t > 0):
    solve()
    t -= 1

File: syllabus-1/day-2/RISK.cpp
#include <bits/stdc++.h>
using namespace std;

#define endl "\n"
#define ll long long

vector<pair<int, int>> cord = { {-1,0},{1,0},{0,-1},{0,1} };
bool isValid(int x, int y, int n, int m) {
    return (x >= 0 && y >= 0 && x < m&& y < n);
}

void solve() {
    int n, m;
    cin >> n >> m;

    string arr[n];
    for (int i = 0;i < n;++i) {
        cin >> arr[i];
    }

    vector<int> res;
    int count = 0;
    for (int i = 0;i < n;++i) {
        for (int j = 0;j < m;++j) {
            // do bfs only when land is there i.e. 1
            if (arr[i][j] == '0') continue;

            // pair is for {i,j}
            queue<pair<int, int>> q;
            q.push({ j,i });
            arr[i][j] = '0';

            while (q.empty() == false) {
                int len = q.size();
                while (len--) {
                    pair<int, int> temp = q.front();
                    q.pop();

                    int x = temp.first;
                    int y = temp.second;

                    count++;

                    for (int k = 0;k < 4;++k) {
                        int cx = x + cord[k].first;
                        int cy = y + cord[k].second;

                        if (isValid(cx, cy, n, m) && arr[cy][cx] == '1') {
                            arr[cy][cx] = '0';
                            q.push({ cx,cy });
                        }
                    }
                }
            }

            // for(string s : arr){
            //     cout<<s<<endl;
            // }cout<<endl;

            res.push_back(count);
            count = 0;
        }
    }

    sort(res.rbegin(), res.rend());

    int ans = 0;
    for (int i = 1;i < res.size();i += 2) {
        ans += res[i];
    }
    cout << ans << endl;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);cout.tie(NULL);

    int t;
    cin >> t;
    while (t--) {
        solve();
    }

    return 0;
}